### PR TITLE
feat(subtreevalidation): skip subtreeData download for blocks whose txs we already have

### PR DESCRIPTION
**Pull request by freemans13**
## Summary
- Adds \`SubtreeValidation.AssumeTxsBroadcastToAllNodes\` (default **off**) and \`LivenessWindow\` (default **5m**) settings.
- Stamps header first-seen time in the blockchain service. Two RPCs surface this:
  - \`ReportPeerBlockHeaderSeen(hash)\` — called from \`blockvalidation.BlockFound\` as soon as a peer-announced header arrives (only when the gate setting is on, so default-off nodes pay zero extra RPC cost).
  - \`GetHeaderReceivedAt(hash) → (time, found, error)\` — consulted by the gate. \`ClientI\` surfaces both methods for in-service callers.
- In-memory \`receivedAt\` store TTL derived from \`LivenessWindow\` via \`max(30m, 2*LivenessWindow)\` so operator-configurable windows can never outlive the store.
- Gate is consulted **once per block** and **never consults FSM state** (see FSM regression test).
- Two call sites gated — both skip \`subtreeData\` download when the block's header was received within the liveness window:
  - \`services/subtreevalidation/check_block_subtrees.go\` — per-subtree workers short-circuit before the \`subtreeStore.Exists(..., FileTypeSubtreeData)\` check. In that path \`subtreeTxs[subtreeIdx]\` stays empty so the block-level batch \`processTransactionsInLevels\` call is skipped entirely (\`batchTxCount == 0\`); tx resolution happens downstream via manifest + UTXO/TxMetaCache + per-tx fetch fallback.
  - \`services/blockvalidation/get_blocks.go\` — \`blockWorker\` skips \`fetchSubtreeDataForBlock\` during catchup.
- One new Prometheus counter: \`teranode_subtreevalidation_livenessgate_decision_total{decision=subtreeonly|subtreedata|notfound|err}\`.

## Shared decision helper

The \`services/subtreevalidation/livenessgate\` package exposes a pure \`Decide(ctx, client, hash, enabled, window) → (Decision, error)\` function consumed by both call sites. The \`Decision\` enum (\`DecisionSubtreeOnly / DecisionSubtreeData / DecisionNotFound / DecisionError\`) drives both the boolean check at the call site and the Prometheus metric label. The \`(*Server).ShouldUseSubtreeOnlyPath\` method in subtreevalidation wraps \`Decide\` with metric emission and debug logging.

The package sits under subtreevalidation because it is a feature-specific decision helper for the subtree-only path, not a general-purpose utility.

## Design

Key invariant: the gate reads only the wall-clock age of the block header's \`ReceivedAt\` stamp. It does **not** consult FSM state — that's the specific behaviour that caused PR #598 to cascade under load and forced its revert via PR #647. A dedicated regression test (\`TestCheckBlockSubtrees_FSMCascadeDoesNotPoisonGate\`) locks this in.

## Test plan

- [x] Unit tests — \`receivedAtStore\` semantics (stamp-and-lookup, write-once, absent, concurrent writes, TTL expiration) — 5 tests.
- [x] Unit tests — \`receivedAtTTL\` helper (floor + scaled branches).
- [x] Unit tests — \`livenessgate.Decide\` truth table — 5 cases.
- [x] Unit tests — \`(*Server).ShouldUseSubtreeOnlyPath\` (metrics-wrapped) truth table — 8 cases.
- [x] Integration tests — \`CheckBlockSubtrees\` gated path: setting off, fresh stamp + skip, stale stamp + fetch — 3 sub-tests.
- [x] Integration tests — \`blockWorker\` gated pre-fetch: setting off, fresh stamp, stale stamp — 3 sub-tests.
- [x] Integration tests — \`BlockFound\` only calls \`ReportPeerBlockHeaderSeen\` when the gate setting is enabled.
- [x] FSM cascade regression guard — forces CATCHINGBLOCKS while fresh stamp exists; gate still fires.
- [x] \`make lint\` — 0 issues.
- [x] Full test sweep for touched packages passes (1707 tests).
- [ ] Dev-scale manual test: enable on dev-scale-1 / dev-scale-2 via ArgoCD, observe \`livenessgate_decision_total{decision="subtreeonly"}\` dominates for ≥24h.

## Rollout

Default \`false\` means merging is a no-op for mainnet / testnet / teratestnet. Enable via \`settings_local.conf\` on dev-scale only.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: services/subtreevalidation/livenessgate.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+		prometheusLivenessGateDecision.WithLabelValues("notfound").Inc()
+		return false
+	}
+	if time.Since(stamp) > u.settings.SubtreeValidation.LivenessWindow {
+		prometheusLivenessGateDecision.WithLabelValues("subtreedata").Inc()
+		return false
+	}
+
+	prometheusLivenessGateDecision.WithLabelValues("subtreeonly").Inc()
+	return true
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/subtreevalidation/livenessgate.go]: Good point. Refactored to push the decision logic into a single `livenessgate.Decide` helper that returns a `Decision` enum (with stable string labels for metrics). The subtreevalidation wrapper now delegates the decision and only layers the Prometheus counter + debug log on top, so both callers (per-subtree check + catchup prefetch) and any future sites will share exactly one definition of "live." Added a test for `Decide` covering all branches.

File: services/subtreevalidation/livenessgate_test.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		wantResult bool
+	}{
+		{"setting off, fresh stamp", false, time.Minute, now.Add(-10 * time.Second), true, nil, false},
+		{"setting off, no stamp", false, time.Minute, time.Time{}, false, nil, false},
+		{"setting on, fresh stamp within window", true, time.Minute, now.Add(-10 * time.Second), true, nil, true},
+		{"setting on, stale stamp outside window", true, time.Minute, now.Add(-5 * time.Minute), true, nil, false},
+		{"setting on, stamp absent", true, time.Minute, time.Time{}, false, nil, false},
+		{"setting on, lookup error", true, time.Minute, time.Time{}, false, errors.NewError("rpc"), false},
+		{"setting on, stamp exactly at window boundary", true, time.Minute, now.Add(-59 * time.Second), true, nil, true},
+		{"setting on, stamp past window boundary by 1s", true, time.Minute, now.Add(-61 * time.Second), true, nil, false},
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/subtreevalidation/livenessgate_test.go]: Agreed — applied the suggested widening (30s vs 90s for the 60s window). I kept the existing boundary cases too; in practice Go's `time.Since` granularity plus test scheduling jitter has never flipped them in local runs, but the wider margins guarantee safety under slow/loaded CI. An injected clock would be even better but felt heavier than warranted for a pure-helper-plus-thin-wrapper; happy to refactor if it flakes again.

File: services/blockvalidation/get_blocks.go [hunk 3]
```diff
@@ -208,6 +209,31 @@ func (u *Server) blockWorker(ctx context.Context, workerID int, workQueue <-chan
...
+					block: work.block,
+					index: work.index,
+				}
+				select {
+				case resultQueue <- result:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockvalidation/get_blocks.go:234]: Real bug — good catch. For the live (non-catchup) path the stamp wasn't written until `AddBlock`, which happens *after* `CheckBlockSubtrees`, so the gate saw `notfound` and the optimization was effectively a no-op. 

Fixed by adding a `ReportPeerBlockHeaderSeen` RPC on the blockchain service (Interface.go / Client.go / LocalClient.go / Server.go / mock.go, plus the proto) and calling it from `blockvalidation.BlockFound` as soon as the hash is parsed — that's the earliest point we know about a peer-announced live block. Stamping errors are informational only (debug-logged, validation continues) since a missed stamp just degrades to the existing subtreeData path.

For the catchup prefetch site (this file), the blocks being pre-fetched are historical by definition, so the stamp correctly remains absent and the gate falls back to subtreeData — which is the intended behaviour there.

File: services/blockchain/Server.go [hunk 4]
```diff
@@ -843,6 +848,10 @@ func (b *Blockchain) AddBlock(ctx context.Context, request *blockchain_api.AddBl
 		return nil, errors.WrapGRPC(err)
 	}
 
+	// Record first-seen time for the subtree-only liveness gate.
+	// See docs/superpowers/specs/2026-04-16-subtree-only-validation-with-liveness-gate-design.md.
+	b.receivedAt.stamp(block.Hash())
+
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockchain/Server.go]: Agreed — the `AddBlock` stamp is kept as a belt-and-braces fallback but is no longer the primary write path. Added a new `ReportPeerBlockHeaderSeen` RPC (Interface.go / Client.go / LocalClient.go / Server.go / mock.go / proto regenerated) that `blockvalidation.BlockFound` now calls as soon as the hash is parsed — that's the earliest moment we learn about a peer-announced live block, well before the subtree fetch / `CheckBlockSubtrees` decisions. This makes the gate actually fire in production.

File: services/blockchain/Server.go [hunk 5]
```diff
@@ -163,6 +167,7 @@ func New(ctx context.Context, logger ulogger.Logger, tSettings *settings.Setting
 		AppCtx:                        ctx,
 		blocksFinalKafkaAsyncProducer: blocksFinalKafkaAsyncProducer,
 		batchTokens:                   make(map[string]*blobDeletionBatchToken),
+		receivedAt:                    newReceivedAtStore(30 * time.Minute),
 	}
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockchain/Server.go]: Fixed — TTL is now derived from `SubtreeValidation.LivenessWindow` via a small helper `receivedAtTTL()` that returns `max(30m, 2*LivenessWindow)` (the 30m floor guards against tiny misconfigured windows shrinking the store so aggressively that stamps expire before any gate check; the 2× multiplier gives headroom for clock drift and operator window tweaks). Applied at both `Server.New` and `NewLocalClient` construction sites with nil-settings-safe fallbacks. Added a `TestReceivedAtTTL` covering the floor, the crossover point, and scaled cases.

File: util/livenessgate/gate.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
+// Package livenessgate provides a shared, pure decision function used by
+// block validation and subtree validation to decide whether a block can
+// use the subtree-only path (skipping subtreeData downloads) based on
+// the age of its header's first-seen timestamp.
+//
+// See docs/superpowers/specs/2026-04-16-subtree-only-validation-with-liveness-gate-design.md.
```
[the remark this answers is not in the document]
Comment (freemans13) [reply util/livenessgate/gate.go]: Fair point — that spec path is gitignored in this repo (noted in the PR description), so the inline references just dangle. Removed the `// See docs/superpowers/specs/...` references from every Go file and replaced the package doc with an inline explanation of what the gate does and when it fires. The design rationale now lives entirely in the PR description instead of as broken paths in source.

File: services/subtreevalidation/check_block_subtrees.go [hunk 7]
```diff
@@ -262,6 +269,15 @@ func (u *Server) CheckBlockSubtrees(ctx context.Context, request *subtreevalidat
 				// PHASE 2: Exact pre-allocation
 				subtreeTxs[subtreeIdx] = make([]*bt.Tx, 0, subtreeToCheck.Length())
 
+				if subtreeOnly {
+					// Gate is on and the block's header was received recently — skip
+					// the subtreeData download entirely. The subtree structure was
+					// already populated from the hash manifest above; tx resolution
+					// happens later in processTransactionsInLevels via UTXO/TxMetaCache,
+					// with per-tx fetch for any real misses.
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/subtreevalidation/check_block_subtrees.go]: You're right — the comment described the non-gated code path. Updated to reflect reality: in the gated path subtreeTxs stays empty, batchTxCount is 0, and processTransactionsInLevels never runs; tx resolution happens downstream in validation via the subtree manifest + UTXO/TxMetaCache, with per-tx fetch as the fallback for real misses.

File: util/kafka/kafka_message/kafka_messages.pb.go [hunk 8]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
 // 	protoc-gen-go v1.36.11
-// 	protoc        v7.34.1
+// 	protoc        v6.33.4
 // source: util/kafka/kafka_message/kafka_messages.proto
```
[the remark this answers is not in the document]
Comment (freemans13) [reply util/kafka/kafka_message/kafka_messages.pb.go:5]: Reverted the `protoc v7.34.1` → `v6.33.4` header-only churn on the 23 unrelated `*.pb.go` files (`git checkout upstream/main -- <paths>`). The only remaining `*.pb.go` diffs are in `services/blockchain/blockchain_api/` where this PR actually adds a new RPC (`ReportPeerBlockHeaderSeen`). Agree on pinning the toolchain — will raise that separately so everyone produces stable outputs going forward.

File: services/blockchain/blockchain_api/blockchain_api_grpc.pb.go [hunk 9]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go-grpc. DO NOT EDIT.
 // versions:
 // - protoc-gen-go-grpc v1.6.1
-// - protoc             v7.34.1
+// - protoc             v6.33.4
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockchain/blockchain_api/blockchain_api_grpc.pb.go]: Fixed — patched the `protoc` header in the two regenerated blockchain_api files from `v6.33.4` back to `v7.34.1` to match the rest of the repo. The stub content is identical (same fields, same service descriptor); only the version-comment lines change. Agreed on pinning the toolchain to stop this happening; I'll raise that as its own issue/PR so the fix is visible to everyone, not buried in a feature PR.

File: services/blockvalidation/get_blocks_test.go [hunk 10]
```diff
@@ -3193,3 +3195,172 @@ func TestFetchBlocksConcurrently_BlockHeightIsSet(t *testing.T) {
...
+			settings:         settings,
+			subtreeStore:     memory.New(),
+			blockchainClient: mockBC,
+			stats:            gocore.NewStat("livenessgate_test"),
+		}, mockBC
+	}
+
+	// A minimal block that has one subtree (so fetchSubtreeDataForBlock has work to do).
+	subtreeHash := createTestHash("liveness-gate-subtree")
+	block := &model.Block{
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockvalidation/get_blocks_test.go:3237]: Great catch — confirmed the panic locally (`SIGSEGV` inside `BlockHeader.Bytes -> HashPrevBlock.CloneBytes`). Populated the test block header with non-nil `HashPrevBlock`, `HashMerkleRoot`, a zero `NBit`, and a timestamp so `block.Hash()` serializes cleanly. All three sub-cases (`SettingOff`, `SettingOn_FreshStamp`, `SettingOn_StaleStamp`) now run and pass.

File: services/subtreevalidation/metrics.go [hunk 11]
```diff
@@ -77,6 +77,15 @@ var (
...
 	prometheusSubtreeValidationSetTXMetaCacheKafkaErrors prometheus.Counter
+
+	// prometheusLivenessGateDecision counts liveness-gate outcomes, labelled by decision.
+	// Labels: subtreeonly, subtreedata, notfound, err.
+	prometheusLivenessGateDecision *prometheus.CounterVec
+
+	// prometheusSubtreeOnlyTxMiss counts per-tx fetch fallbacks during the subtree-only
+	// path. A spike on a node configured with AssumeTxsBroadcastToAllNodes indicates
+	// upstream trouble (blaster health, peer connectivity, or misconfiguration).
+	prometheusSubtreeOnlyTxMiss prometheus.Counter
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/subtreevalidation/metrics.go]: Good point — the metric was forward-looking (the PR description mentions it, but the wiring into the per-tx fetch path in the validator didn't land with this PR). Rather than leave a silent zero that could mislead dashboards, I've removed the declaration and the `promauto.NewCounter` registration. When the per-tx fetch instrumentation is added (separate PR), the metric can come back alongside its increment site. PR description will be updated in a follow-up edit to reflect the single metric this PR actually emits.

File: services/blockvalidation/Server.go [hunk 12]
```diff
@@ -1122,6 +1122,15 @@ func (u *Server) BlockFound(ctx context.Context, req *blockvalidation_api.BlockF
 			errors.NewProcessingError("[BlockFound][%s] failed to create hash from bytes", util.ReverseAndHexEncodeSlice(req.Hash), err))
 	}
 
+	// Seed the subtree-only liveness gate with a first-seen timestamp for this
+	// header. Must happen BEFORE CheckBlockSubtrees / subtreeData fetch decisions
+	// otherwise the gate only ever sees stamps written after AddBlock, making the
+	// optimisation a no-op in production.
+	// Stamping errors are informational — do not fail validation on a stamp failure.
+	if stampErr := u.blockchainClient.ReportPeerBlockHeaderSeen(ctx, hash); stampErr != nil {
+		u.logger.Debugf("[BlockFound][%s] ReportPeerBlockHeaderSeen failed, liveness gate may miss for this header: %v", hash.String(), stampErr)
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockvalidation/Server.go]: Good catch — gated the stamp call behind `AssumeTxsBroadcastToAllNodes && LivenessWindow > 0` (matching your suggestion). Nodes running with the default (mainnet / testnet / teratestnet) now skip the RPC entirely, keeping `BlockFound` zero-overhead as promised. Also added a subtest (`ReportPeerBlockHeaderSeen only fires when gate setting is enabled`) with `AssertNotCalled` to lock that in.

File: services/subtreevalidation/metrics.go [hunk 13]
```diff
@@ -185,4 +189,14 @@ func _initPrometheusMetrics() {
...
+
+	prometheusLivenessGateDecision = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "teranode",
+			Subsystem: "subtreevalidation",
+			Name:      "livenessgate_decision_total",
+			Help:      "Count of liveness-gate decisions by outcome (subtreeonly, subtreedata, notfound, err).",
+		},
+		[]string{"decision"},
+	)
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/subtreevalidation/metrics.go:201]: Done — updated the PR description to mention only the single `livenessgate_decision_total` counter this PR actually emits. The `subtreeonly_tx_miss_total` counter was forward-looking in the original draft; I've noted it as deferred until the per-tx fetch instrumentation lands in a follow-up.

File: util/livenessgate/gate.go [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+//
+// The gate resolves to DecisionSubtreeOnly only when BOTH:
+//   - enabled is true (typically from SubtreeValidation.AssumeTxsBroadcastToAllNodes), AND
+//   - the blockchain client has a ReceivedAt stamp for this header newer than window.
+//
+// Any other outcome (disabled, error, absent stamp, stale stamp) resolves
+// to a subtreeData fallback. The gate is an optimization, never a
+// correctness constraint.
+func Decide(ctx context.Context, client Client, blockHash *chainhash.Hash, enabled bool, window time.Duration) (Decision, error) {
+	if !enabled {
```
[the remark this answers is not in the document]
Comment (freemans13) [reply util/livenessgate/gate.go]: Good consistency fix. `Decide` now short-circuits on `!enabled || window <= 0` before the client call, so `AssumeTxsBroadcastToAllNodes=true` with `LivenessWindow=0` (or negative) is treated identically to the setting being off — no RPC, no metrics noise from "err" or "notfound" buckets. Also extended the `Decide` truth table with three new cases (`disabled`, `zero window`, `negative window`) that all assert `sc.getCalls == 0` to lock in that the client is never touched.

File: services/blockvalidation/Server.go [hunk 15]
```diff
@@ -1122,6 +1122,21 @@ func (u *Server) BlockFound(ctx context.Context, req *blockvalidation_api.BlockF
...
+	//
+	// Gated behind the gate setting itself so nodes that do not use the
+	// optimisation (mainnet/testnet default) don't pay the extra blockchain-service
+	// RPC on the BlockFound hot path. Stamping errors are informational — do not
+	// fail validation on a stamp failure.
+	if u.settings.SubtreeValidation.AssumeTxsBroadcastToAllNodes && u.settings.SubtreeValidation.LivenessWindow > 0 {
+		if stampErr := u.blockchainClient.ReportPeerBlockHeaderSeen(ctx, hash); stampErr != nil {
+			u.logger.Debugf("[BlockFound][%s] ReportPeerBlockHeaderSeen failed, liveness gate may miss for this header: %v", hash.String(), stampErr)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockvalidation/Server.go]: Agreed. Moved the stamp call to after the `GetBlockExists` check and gated it on `exists == false` (via early return). Duplicate `BlockFound` announcements for already-validated blocks now return immediately without the extra blockchain-service RPC, while new blocks still get stamped before going on the `blockFoundCh` queue (so the stamp lands before any subtree fetch decision downstream).

File: services/blockchain/received_at.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+	mu sync.Mutex
+	m  *expiringmap.ExpiringMap[chainhash.Hash, time.Time]
+}
+
+func newReceivedAtStore(ttl time.Duration) *receivedAtStore {
+	return &receivedAtStore{
+		m: expiringmap.New[chainhash.Hash, time.Time](ttl),
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockchain/received_at.go:35]: Good catch — goroutine leak confirmed. Added `(*receivedAtStore).Stop()` that forwards to the underlying `ExpiringMap.Stop()` (nil-safe and idempotent), and wired `(*Blockchain).Stop` to call it so the background cleanup ticker is released on shutdown. That fixes the leak for the primary `Blockchain` instance. LocalClient also owns a `receivedAtStore` but has no lifecycle method on `ClientI` today; adding one would be a broader API change — happy to tackle that in a separate PR if worth it, as the current LocalClient usage is primarily in short-lived test contexts where the test framework tears down the process anyway.

File: services/blockchain/received_at.go [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+// this node. It is consulted by the subtree-only liveness gate to decide whether
+// validation can skip the subtreeData download path.
+//
+// Entries expire after a TTL (set large relative to SubtreeValidation.LivenessWindow
+// so a live-for-the-window header is never evicted prematurely). Expired or absent
+// entries are reported as "not found" — the gate treats that as "not live" and
+// falls back to subtreeData, which is safe.
+//
+// The first stamp for a given hash wins; repeated inserts are no-ops. This matches
+// the semantic "when did we first learn about this header?"
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockchain/received_at.go]: Right — updated both the struct comment and the `stamp` method's doc to make the TTL boundary explicit: "within a single TTL window the first stamp wins; once the TTL elapses the entry has been evicted and a new stamp records the new observation rather than reviving a stale one." In practice a growing chain won't see the same hash twice across a TTL boundary, but the doc now describes the real semantics rather than implying process-lifetime write-once.

File: services/blockchain/LocalClient.go [hunk 18]
```diff
@@ -37,12 +42,20 @@ type LocalClient struct {
...
+
 	return &LocalClient{
 		logger:       logger,
 		settings:     tSettings,
 		store:        store,
 		subtreeStore: subtreeStore,
 		utxoStore:    utxoStore,
+		receivedAt:   newReceivedAtStore(receivedAtTTL(livenessWindow)),
 		subscribers:  make(map[string]chan *blockchain_api.Notification),
 	}, nil
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockchain/LocalClient.go:60]: Added `(*LocalClient).Stop()` (nil-safe, idempotent) that forwards to `receivedAt.Stop()` so transient `LocalClient`s in tests/benchmarks can `defer client.Stop()` without leaking the background cleanup goroutine. Kept it off the `ClientI` interface for now — introducing a lifecycle method there is a broader surface change and none of the other `ClientI` owners need it yet; happy to promote it once we see a second need.

File: services/blockvalidation/Server_test.go [hunk 19]
```diff
@@ -1302,6 +1308,46 @@ func Test_BlockFound(t *testing.T) {
...
+		gateOffSettings.SubtreeValidation.AssumeTxsBroadcastToAllNodes = false
+
+		bv := &BlockValidation{
+			blockHashesCurrentlyValidated: txmap.NewSwissMap(0),
+			blocksCurrentlyValidating:     txmap.NewSyncedMap[chainhash.Hash, *validationResult](),
+			blockExistsCache:              expiringmap.New[chainhash.Hash, bool](120 * time.Minute),
+			blockchainClient:              mockBlockchainClient,
+		}
+		defer bv.blockExistsCache.Stop()
+		require.NoError(t, bv.SetBlockExists(&hash)) // short-circuits after stamp check
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockvalidation/Server_test.go]: Good catch — updated the comment. Also noticed the test had a design issue after the stamp-timing change: it was relying on BOTH `exists=true` AND `gate-off` to skip the stamp, so the assertion was trivially true regardless of the gate guard. Refactored the test to use `exists=false` (so the gate-off guard is the only reason the stamp is skipped) with a drain-and-assert pattern for the queued work. The assertion now actually tests what the name says.

File: services/blockvalidation/Server_test.go [hunk 20]
```diff
@@ -1302,6 +1308,56 @@ func Test_BlockFound(t *testing.T) {
...
+
+		resp, err := server.BlockFound(ctx, req)
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+
+		// Drain the queued work so the background goroutine finishes before we
+		// assert, avoiding a race with the deferred cache.Stop().
+		time.Sleep(10 * time.Millisecond)
+		require.Equal(t, 1, len(server.blockFoundCh))
+		<-server.blockFoundCh
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockvalidation/Server_test.go]: Agreed — replaced the sleep-and-check-len pattern with `select { case <-server.blockFoundCh: ... case <-time.After(time.Second): t.Fatal(...) }` as suggested. The test now blocks deterministically until the background goroutine enqueues, with a generous timeout as the safety net; no more wall-clock gamble.

File: services/subtreevalidation/livenessgate_test.go [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+			tSettings := settings.NewSettings()
+			tSettings.SubtreeValidation.AssumeTxsBroadcastToAllNodes = tc.enabled
+			tSettings.SubtreeValidation.LivenessWindow = tc.window
+
+			srv := &Server{
+				logger:           ulogger.TestLogger{},
+				settings:         tSettings,
+				blockchainClient: mockClient,
+			}
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/subtreevalidation/livenessgate_test.go:62]: Great catch — confirmed this test passes today only because `SubtreeValidation_test.go` runs the package-level `sync.Once` first and leaves the counters registered as a side effect. Added an explicit `InitPrometheusMetrics()` call at the top of `TestShouldUseSubtreeOnlyPath_TruthTable` (it's idempotent via `sync.Once`). The test now works in isolation (`go test -run TestShouldUseSubtreeOnlyPath_TruthTable`) and is order-independent.

File: services/subtreevalidation/check_block_subtrees.go [hunk 22]
```diff
@@ -262,6 +271,17 @@ func (u *Server) CheckBlockSubtrees(ctx context.Context, request *subtreevalidat
...
 
+				if subtreeOnly {
+					// Gate is on and the block's header was received recently — skip
+					// the subtreeData download entirely. subtreeTxs[subtreeIdx] stays
+					// empty, so the block-level batch processTransactionsInLevels call
+					// downstream is skipped (batchTxCount == 0). Transaction resolution
+					// happens later in validation via the subtree manifest +
+					// UTXO/TxMetaCache, with per-tx fetch as the fallback for real
+					// misses.
+					return nil
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/subtreevalidation/check_block_subtrees.go]: Right — moved the `subtreeOnly` early-return ahead of the `make([]*bt.Tx, 0, subtreeToCheck.Length())` pre-allocation. On the gated path `subtreeTxs[subtreeIdx]` now stays as the zero-value nil slice, saving `subtreeToCheck.Length()` capacity worth of memory per missing subtree. The comment downstream still holds (nil slice → `batchTxCount == 0` → `processTransactionsInLevels` skipped).

File: services/blockchain/Server.go [hunk 23]
```diff
@@ -163,6 +174,7 @@ func New(ctx context.Context, logger ulogger.Logger, tSettings *settings.Setting
 		AppCtx:                        ctx,
 		blocksFinalKafkaAsyncProducer: blocksFinalKafkaAsyncProducer,
 		batchTokens:                   make(map[string]*blobDeletionBatchToken),
+		receivedAt:                    newReceivedAtStore(receivedAtTTL(livenessWindow)),
 	}
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockchain/Server.go:183]: Done — the `receivedAt` store is now only allocated when the gate is enabled at construction (`AssumeTxsBroadcastToAllNodes && LivenessWindow > 0`), in both `Server.New` and `NewLocalClient`. Made `receivedAtStore.stamp` and `.lookup` nil-safe so call sites don't need extra guards beyond the existing gate-setting checks they already perform. Default-off deployments now allocate zero bytes and start zero goroutines for this feature — true no-op.

File: services/blockchain/Server.go [hunk 24]
```diff
@@ -843,6 +859,14 @@ func (b *Blockchain) AddBlock(ctx context.Context, request *blockchain_api.AddBl
 		return nil, errors.WrapGRPC(err)
 	}
 
+	// Record first-seen time for the subtree-only liveness gate as a
+	// belt-and-braces fallback. The primary stamp happens earlier in
+	// blockvalidation.BlockFound via ReportPeerBlockHeaderSeen, but this write
+	// covers paths that reach AddBlock without having gone through BlockFound
+	// (e.g. locally mined blocks). Stamping is write-once, so double writes are
+	// cheap no-ops.
+	b.receivedAt.stamp(block.Hash())
```
[the remark this answers is not in the document]
Comment (freemans13) [reply services/blockchain/Server.go]: Agreed — gated the `AddBlock` stamp behind `AssumeTxsBroadcastToAllNodes && LivenessWindow > 0` in both `Server.AddBlock` and `LocalClient.AddBlock`. Combined with the conditional store allocation (other thread), nodes that don't use the optimisation now allocate zero receivedAt entries during fast catchup. The stamp still fires for locally mined blocks on nodes where the gate is enabled (they reach AddBlock without going through BlockFound), which is the original belt-and-braces intent.

**Comment by freemans13** [comment]
Superseded by #745 — self-discovering adaptive gate replaces the operator-configured feature flag + wall-clock liveness gate. The new design deliberately avoids any wall-clock or FSM-state gating (enforced by a source-scanning regression test), eliminating the class of bug that caused PR #598 to be reverted via #647.
